File: P53 Maximum Subarray/Solution.java
class Solution {
    public int maxSubArray(int[] nums) {
        // Kadane’s Algorithm
        // Time complexity: O(n)
        int result = Integer.MIN_VALUE, maxTemp = 0;
        for(int i = 0; i < nums.length; i++) {
            maxTemp += nums[i];
            if(result < maxTemp) {
                result = maxTemp;
            }
            if(maxTemp < 0) {
                maxTemp = 0;
            }
        }
        return result;
    }

    public int maxSubArray2(int[] nums) {
        // divide and conquer
        // Time complexity: O(nlogn)
        return maxSubArraySum(nums, 0, nums.length-1);
    }

    public int maxSubArraySum(int nums[], int l, int h) {
        if (l == h) return nums[l];

        int m = (l + h)/2;

        // T(n) = 2T(n/2) + Θ(n)
        return Math.max(Math.max(maxSubArraySum(nums, l, m), maxSubArraySum(nums, m+1, h)),
                maxCrossingSum(nums, l, m, h));
    }

    public int maxCrossingSum(int nums[], int l, int m, int h) {
        int sum = 0;
        int leftSum = Integer.MIN_VALUE;
        for (int i = l; i <= m; i++) {
            sum = sum + nums[i];
            if (sum > leftSum) {
                leftSum = sum;
            }
        }

        sum = 0;
        int rightSum = Integer.MIN_VALUE;
        for (int i = m + 1; i <= h; i++) {
            sum = sum + nums[i];
            if (sum > rightSum) {
                rightSum = sum;
            }
        }

        return leftSum + rightSum;
    }

}

File: P53 Maximum Subarray/Solution.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Kadane’s Algorithm
        # Time complexity: O(n)
        result = -sys.maxint -1
        maxTemp = 0
        for num in nums:
            maxTemp += num
            if result < maxTemp:
                result = maxTemp
            if maxTemp < 0:
                maxTemp = 0
        return result

    def maxSubArray2(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # divide and conquer
        # Time complexity: O(nlogn)
        return self.maxSubArraySum(nums, 0, len(nums)-1)

    def maxSubArraySum(self, nums, l, h):
        if l == h:
            return nums[l]

        m = (l+ h) // 2

        return max(self.maxSubArraySum(nums, l, m), self.maxSubArraySum(nums, m+1, h), self.maxCrossingSum(nums, l, m, h))

    def maxCrossingSum(self, nums, l, m, h):
        sum = 0
        left_sum = -sys.maxint -1
        for i in range(m, l-1, -1):
            sum += nums[i]
            if sum > left_sum:
                left_sum = sum

        sum = 0
        right_sum = -sys.maxint -1
        for i in range(m+1, h+1):
            sum += nums[i]
            if sum > right_sum:
                right_sum = sum

        return left_sum + right_sum